emails/subscriptions.go

```
@@ -10,46 +10,111 @@ import (
+		user, err := queries.GetUserById(c, input.UserID)
+		if err != nil {
+			util.ErrResponse(c, http.StatusInternalServerError, err)
+			return
+		}
+
+		if user.Email == "" {
+			util.ErrResponse(c, http.StatusBadRequest, errNoEmailSet{input.UserID})
 			return
 		}
-		userID := input.UserID
-		if input.JWT != "" {
-			userID, err = auth.JWTParse(input.JWT, viper.GetString("JWT_SECRET"))
-			if err != nil {
-				util.ErrResponse(c, http.StatusBadRequest, err)
+
+		errGroup := new(errgroup.Group)
+
+		for _, emailType := range emailTypes {
+			switch emailType {
+			case model.EmailUnsubscriptionTypeAll:
+				errGroup.Go(func() error {
+					if input.Unsubs.All {
+						return addEmailToGlobalUnsubscribeGroup(c, user.Email.String())
+					}
+					return removeEmailFromGlobalUnsubscribeGroup(c, user.Email.String())
+				})
+
+			case model.EmailUnsubscriptionTypeNotifications:
+
+				errGroup.Go(func() error {
+					if input.Unsubs.Notifications {
+						return addEmailToUnsubscribeGroup(c, user.Email.String(), viper.GetString("SENDGRID_UNSUBSCRIBE_NOTIFICATIONS_GROUP_ID"))
+					}
+					return removeEmailFromUnsubscribeGroup(c, user.Email.String(), viper.GetString("SENDGRID_UNSUBSCRIBE_NOTIFICATIONS_GROUP_ID"))
+				})
+			default:
+				util.ErrResponse(c, http.StatusBadRequest, fmt.Errorf("unsupported email type: %s", emailType))
 				return
 			}
+
+		}
+
+		logger.For(c).Infof("unsubscribing user %s from email types: %+v", input.UserID, input.Unsubs)
+		err = queries.UpdateUserEmailUnsubscriptions(c, coredb.UpdateUserEmailUnsubscriptionsParams{
+			ID:                   input.UserID,
+			EmailUnsubscriptions: input.Unsubs,
+		})
+		if err != nil {
+			util.ErrResponse(c, http.StatusInternalServerError, err)
+			return
+		}
+
+		if err := errGroup.Wait(); err != nil {
+			logger.For(c).Errorf("error unsubscribing user %s from email types %+v: %s", input.UserID, input.Unsubs, err)
```

Should we return an error code here or its ok to return a 200?

---

It is ok, if sendgrid does not update we don't need to fail because sendgrid will fail when a user is already on a list for example. It seems like we just want to **try** to keep in sync with sendgrid. If we do want to return an error here, we just have to make sure it does not bubble to the frontend because it is still not something the frontend should handle.